cpp/src/arrow/util/string_header.h

```diff
@@ -0,0 +1,377 @@
+    if (PrefixAsInt() != other.PrefixAsInt()) {
+      // The result is decided on prefix. The shorter will be less
+      // because the prefix is padded with zeros.
+      return memcmp(prefix_.data(), other.prefix_.data(), kPrefixSize);
+    }
+    int32_t size = std::min(size_, other.size_) - kPrefixSize;
+    if (size <= 0) {
+      // One string is just the prefix.
+      return size_ - other.size_;
+    }
+    if (static_cast<uint32_t>(size) <= kInlineSize && IsInline() && other.IsInline()) {
+      int32_t result = memcmp(value_.inlined.data(), other.value_.inlined.data(), size);
+      return (result != 0) ? result : size_ - other.size_;
+    }
+    int32_t result = memcmp(data() + kPrefixSize, other.data() + kPrefixSize, size);
+    return (result != 0) ? result : size_ - other.size_;
+  }
+
+  template <typename BufferPtr>
+  int CompareIndexOffset(const BufferPtr* char_buffers, const StringHeader& other,
+                         const BufferPtr* other_char_buffers) const {
+    if (PrefixAsInt() != other.PrefixAsInt()) {
+      // The result is decided on prefix. The shorter will be less
+      // because the prefix is padded with zeros.
+      return memcmp(prefix_.data(), other.prefix_.data(), kPrefixSize);
+    }
+    int32_t size = std::min(size_, other.size_) - kPrefixSize;
+    if (size <= 0) {
+      // One string is just the prefix.
+      return size_ - other.size_;
+    }
+    if (static_cast<uint32_t>(size) <= kInlineSize && IsInline() && other.IsInline()) {
+      int32_t result = memcmp(value_.inlined.data(), other.value_.inlined.data(), size);
+      return (result != 0) ? result : size_ - other.size_;
+    }
+
+    int32_t result = memcmp(
+        GetPointerFromBuffersOrInlineData(char_buffers) + kPrefixSize,
+        other.GetPointerFromBuffersOrInlineData(other_char_buffers) + kPrefixSize, size);
+    return (result != 0) ? result : size_ - other.size_;
+  }
+
+  int64_t SizeAndPrefixAsInt64() const {
+    return reinterpret_cast<const int64_t*>(this)[0];
+  }
+
+  int64_t InlinedAsInt64() const { return reinterpret_cast<const int64_t*>(this)[1]; }
+
+  int32_t PrefixAsInt() const { return *reinterpret_cast<const int32_t*>(&prefix_); }
+
+  // FIXME(bkietz) replace this with a std::array<uint8_t, 16> and forgo the union.
+  // Type punning (AKA violation of the strict aliasing rule) is undefined behavior.
+  // Using memcpy to access the bytes of the object representation of trivially copyable
+  // objects is not undefined behavior. Given sufficiently explicit hints on alignment
+  // and size, compilers elide memcpy calls in favor of identical assembly to what
+  // the type punning implementation produces.
+  // We rely on all members being laid out top to bottom . C++
+  // guarantees this.
+  uint32_t size_ = 0;
+  std::array<char, kPrefixSize> prefix_ = {0};
+  union {
+    std::array<char, 8> inlined = {0};
+    const char* data;
+    struct {
+      uint32_t buffer_index;
+      uint32_t offset;
```

Comment:
Why unsigned integers?  Is there some place you are relying on the rollover behavior?  This just seems slightly out of sync with the way strings (as in normal strings, not views) are done (e.g. using `int32_t`).

Reply:
I use unsigned for compatibility with other implementations. It's true that this is a deviation from the way `arrow::` usually does things, but compatibility seemed more important in this case. I call this deviation out in `Columnar.rst`, for what that's worth

Reply:
Well, Java generally favors signed integers, AFAICT.